Add doxygen documentation in API headers (Fixes #80)

a1lu opened the pull request:
Added doxygen docs to the API at include/criterion

Snaipe commented on include/criterion/redirect.h:
```diff
@@ -100,42 +104,441 @@ CR_API CR_STDN FILE *cr_mock_file_size(size_t max_size);
+ * is aborted.
+ * The optional string is printed on failure.
+ *
+ * @param[in] ExpectedContents C string with the ExpectedContents
```
I believe this should be `RefFile` rather than `ExpectedContents`.

Snaipe commented on include/criterion/redirect.h:
```diff
@@ -100,42 +104,441 @@ CR_API CR_STDN FILE *cr_mock_file_size(size_t max_size);
+ * Otherwise the test is marked as failure but the execution will continue.
+ * The optional string is printed on failure.
+ *
+ * @param[in] ExpectedContents C string with the ExpectedContents
```
Same comment for `RefFile`

Snaipe commented on include/criterion/redirect.h:
```diff
@@ -100,42 +104,441 @@ CR_API CR_STDN FILE *cr_mock_file_size(size_t max_size);
+ * is aborted.
+ * The optional string is printed on failure.
+ *
+ * @param[in] UnexpectedContents C string with the UnexpectedContents
```
Same comment for `RefFile`

Snaipe commented on include/criterion/redirect.h:
```diff
@@ -100,42 +104,441 @@ CR_API CR_STDN FILE *cr_mock_file_size(size_t max_size);
+ * Otherwise the test is marked as failure but the execution will continue.
+ * The optional string is printed on failure.
+ *
+ * @param[in] UnexpectedContents C string with the UnexpectedContents
```
Same comment for `RefFile`

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * The test is marked as failure and the execution of the function is aborted.
+ * The optional string is printed on failure.
+ *
+ * @param[in] fmt printf like format string
```
I just realized, but `fmt` should be `FormatString`.

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+/**
+ * Passes if Actual is byte-to-byte equal to Expected
+ *
+ * Call: \c cr_assert_arr_eq(Actual, Expected, [FormatString, [Args...]])\n
```
This should be `cr_expect_arr_eq`

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+/**
+ * Passes if Actual is not byte-to-byte equal to Expected
+ *
+ * Call: \c cr_assert_arr_neq(Actual, Expected, Size, [FormatString,
```
This should be `cr_assert_arr_neq`

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+/**
+ * Passes if Actual is comparatively equal to Expected (C++ / GNU C99 only)
+ *
+ * Call: \c cr_assert_arr_eq_cmp(Actual, Expected, Size, Cmp, [FormatString,
```
This should be `cr_assert_arr_eq_cmp`

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * Call: \c cr_assert_arr_eq(Actual, Expected, [FormatString, [Args...]])\n
+ * Passes if Actual is byte-to-byte equal to Expected.
+ * Otherwise the test is marked as failure but the execution will continue.
+ * The optional string is printed on failure.
```
I think the note from the original docs is missing, adding a warning like "This should not be used on arrays of struct, consider using cr_assert_arr_eq_cmp instead." could be nice.
(same thing for `cr_assert_arr_neq`)

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+/**@}*/
+
+/**
+ * @defgroup SafeArrCmpAsserts Safe array comparison assertions
```
There should definitely be something saying that these assertions are C++ and GNU C only. Once in the group, then once per macro would be nice.

Snaipe commented on include/criterion/assert.h:
```diff
@@ -157,19 +1094,144 @@
+/**@}*/
+
+/**
+ * @defgroup ExceptionAsserts Exception asserts
```
Same as the safe array assertions, there should be a notice that these are C++-only, even thought it might be obvious.

Snaipe commented on include/criterion/abort.h:
```diff
@@ -42,6 +46,11 @@ CR_API CR_NORETURN void criterion_abort_test(void);
+ *  Kills the current test, marking it as failed.
+ *
+ *  @param[in] msg printf like format string
+ */
```
You're missing an `@param[in] ... format parameters` here.

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * The optional string is printed on failure.
+ *
+ * @param[in] fmt printf like format string
+ *
```
All macros should probably have an `@param[in] Args... the format paramaters` or something.

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * Passes if Actual is less or equal to Reference
+ *
+ * Call: \c cr_assert_leq(Actual, Reference, [FormatString, [Args...]])\n
+ * Passes if Actual is less to Reference.
```
Less or equal to

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * Passes if Actual is less or equal to Reference
+ *
+ * Call: \c cr_expect_leq(Actual, Reference, [FormatString, [Args...]])\n
+ * Passes if Actual is less to Reference.
```
Less or equal to

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * Passes if Actual is greater or equal to Reference
+ *
+ * Call: \c cr_assert_geq(Actual, Reference, [FormatString, [Args...]])\n
+ * Passes if Actual is greater to Reference.
```
Greater or equal to

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
+ * Passes if Actual is greater or equal to Reference
+ *
+ * Call: \c cr_expect_geq(Actual, Reference, [FormatString, [Args...]])\n
+ * Passes if Actual is greater to Reference.
```
Greater or equal to

Snaipe commented on include/criterion/assert.h:
```diff
@@ -30,109 +35,1041 @@
 
+
+/**
+ * Passes if Actual is comparatively equal to Expected (C++ / GNU C99 only)
```
The `(C++ / GNU C99 only)` should be a full sentence in a note below `The optional string is printed on failure.`, maybe something like:
```
@note
This macro is only available on C++ and GNU C compilers.
```

Snaipe commented:
I think that the doxygen documentation is a bit too separated from the actual sphinx docs, but this might be better done in a separate issue.

Also, 2 points on the Doxyfile (can't comment inside because there's 2000 lines):

1. exclude `criterion/internal` from the doc generator, otherwise the internals will be picked up by the recursive traversal
2. embedding the README.md might be a bad idea (it renders weird), I would rather have a doxygen-specific main page inside `criterion.h`.

a1lu commented:
* fixed the issues
* added some notes and warnings
* removed recursive flag in doxyfile
* removed mainpage


Doxygen prints no \n at the end of the comment lines. Would it be nice to have the same format in the rendered file as in the source files?

Snaipe commented:
As it stands, internal fields in `struct criterion_test_extra_data` are reported, they should be put in a `@internal` section.

> Doxygen prints no \n at the end of the comment lines. Would it be nice to have the same format in the rendered file as in the source files? 

What do you mean? Do you have a specific example in mind?

P.S: You also missed a few issues.

a1lu commented:
Something like
```
* Otherwise the test is marked as failure and the execution of the function
 * is aborted.
 * The optional string is printed on failure.
```
is printed as 
``
Otherwise the test is marked as failure and the execution of the function is aborted. The optional string is printed on failure.
``
in the html output. So we could add some \n to get the same format. Just a style thing.

Also fixed some Expected -> Unexpected and -> Reference things. 
Your comments above should all be fixed

Snaipe commented:
Forcing a line feed in those case seems improper -- the next statement should be put in its own paragraph.

That is, instead of:

```
 * Call: \c cr_assert(Condition, [FormatString, [Args...]]).\n
 * Evaluates the condition and passes if it is true.
 * Otherwise the test is marked as failure and the execution of the function
 * is aborted.
 * The optional string is printed on failure.
```

do:

```
 * Call: \c cr_assert(Condition, [FormatString, [Args...]]).
 *
 * Evaluates the condition and passes if it is true.
 * Otherwise the test is marked as failure and the execution of the function
 * is aborted.
 *
 * The optional string is printed on failure.
```

a1lu commented:
done.

Do you have any notes for the mainpage?

Snaipe commented:
> Do you have any notes for the mainpage?

Considering that I'll try to merge the doxygen docs with the sphinx docs, there's no need for a mainpage.

Otherwise, everything seems good -- if you don't have anything else to submit, I'll merge this in.
